### PR TITLE
Rewrite kyuubi hive jdbc module based on hive branch-2.3 code base

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Hive code base:
https://github.com/apache/hive/commit/8e7f23f34b2ce7328c9d571a13c336f0c8cdecb6

commit 8e7f23f34b2ce7328c9d571a13c336f0c8cdecb6 (HEAD, origin/branch-2.3)
Author: Neelesh Srinivas Salian <nssalian@users.noreply.github.com>
Date:   Tue Oct 19 12:50:31 2021 -0700

    HIVE-25616: Hive-24741 backport to 2.3 (Neelesh Srinivas Salian reviewed by Vihang Karajgaonkar)

    Closes (#2730)

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiHiveDriver.java [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+import java.util.regex.Pattern;
+
+import org.apache.kyuubi.jdbc.hive.Utils.JdbcConnectionParams;
+
+
+/**
+ * HiveDriver.
+ *
+ */
+public class KyuubiHiveDriver implements Driver {
```
Comment (pan3793) [inline kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiHiveDriver.java:41]: We expose this class to user, please don't change the `package`

File: docs/quick_start/quick_start_with_dbeaver.md [hunk 2]
```diff
@@ -86,7 +86,7 @@ Click next...
 
 ### Edit the Driver
 
-We can set libraries that include the `org.apache.hive.jdbc.HiveDriver` and all of its dependencies.
+We can set libraries that include the `KyuubiHiveDriver` and all of its dependencies.
```
Comment (pan3793) [inline docs/quick_start/quick_start_with_dbeaver.md:89]: `org.apache.kyuubi.jdbc.KyuubiHiveDriver`

File: kyuubi-hive-jdbc/pom.xml [hunk 3]
```diff
@@ -30,6 +30,12 @@
     <name>Kyuubi Project Hive JDBC Client</name>
     <packaging>jar</packaging>
 
+    <properties>
+        <!-- only used in kyuubi-hive-jdbc module -->
+        <junit.version>4.11</junit.version>
+        <mockito-all.version>1.9.5</mockito-all.version>
+    </properties>
```
Comment (turboFei) [inline kyuubi-hive-jdbc/pom.xml:37]: cc @pan3793

File: kyuubi-hive-jdbc/pom.xml [hunk 4]
```diff
@@ -237,13 +243,35 @@
...
         <testOutputDirectory>target/scala-${scala.binary.version}/test-classes</testOutputDirectory>
 
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <skipTests>false</skipTests>
+                </configuration>
+            </plugin>
```
Comment (turboFei) [inline kyuubi-hive-jdbc/pom.xml:273]: enable java tests

File: kyuubi-hive-jdbc/pom.xml [hunk 5]
```diff
@@ -237,13 +243,35 @@
...
+            <artifactId>junit</artifactId>
+            <version>${junit.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-all</artifactId>
+            <version>${mockito-all.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (turboFei) [inline kyuubi-hive-jdbc/pom.xml:259]: only used here

**Comment by turboFei** [comment]
image
should we remove kyuubi-hive-jdbc module from kyuubi-codecov?

**Comment by pan3793** [comment]
I think we should remove. KyuubiHiveJDBCDriver actually do sufficient test on `kyuubi-server` and `kyuubi-spark-sql-engine` modules, the codecov does not reflect the fact.

File: kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpKerberosRequestInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+      // Set the session key token (Base64 encoded) in the headers
+      httpRequest.addHeader(HttpAuthUtils.AUTHORIZATION + ": " +
+        HttpAuthUtils.NEGOTIATE + " ", kerberosAuthHeader);
+    } catch (Exception e) {
+      throw new HttpException(e.getMessage(), e);
+    } finally {
+      kerberosLock.unlock();
+    }
+  }
+}
```
Comment (yaooqinn) [inline kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpKerberosRequestInterceptor.java:72]: nit

File: kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpBasicAuthInterceptor.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    }
+  }
+
+  @Override
+  protected void addHttpAuthHeader(HttpRequest httpRequest, HttpContext httpContext)
+    throws Exception {
+    Header basicAuthHeader = authScheme.authenticate(credentials, httpRequest, httpContext);
+    httpRequest.addHeader(basicAuthHeader);
+  }
+}
```
Comment (yaooqinn) [inline kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/HttpBasicAuthInterceptor.java:54]: nit

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-hive-jdbc/pom.xml (hunk 3 above)
Comment (pan3793) [reply kyuubi-hive-jdbc/pom.xml:37]: Let's put them in root pom.xml L146, and also start with hive.jdbc.

File: kyuubi-hive-jdbc/pom.xml [hunk 8]
```diff
@@ -249,13 +243,35 @@
             <groupId>org.apache.hadoop</groupId>
             <artifactId>hadoop-client-api</artifactId>
         </dependency>
+
+        <!-- unit test -->
```
Comment (pan3793) [inline kyuubi-hive-jdbc/pom.xml:247]: The comment makes little sense here.

**Comment by turboFei** [comment]
have addressed all the comments.

File: kyuubi-hive-jdbc/src/test/java/org/apache/kyuubi/jdbc/hive/TestKyuubiPreparedStatement.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+		when(tGetOperationStatusResp.getStatus()).thenReturn(tStatus_SUCCESS);
+		when(tGetOperationStatusResp.getOperationState()).thenReturn(TOperationState.FINISHED_STATE);
+		when(tGetOperationStatusResp.isSetOperationState()).thenReturn(true);
+		when(tGetOperationStatusResp.isSetOperationCompleted()).thenReturn(true);
+
+		when(client.GetOperationStatus(any(TGetOperationStatusReq.class))).thenReturn(tGetOperationStatusResp);
+		when(client.ExecuteStatement(any(TExecuteStatementReq.class))).thenReturn(tExecStatementResp);
+	}
+
+	@SuppressWarnings("resource")
```
Comment (pan3793) [inline kyuubi-hive-jdbc/src/test/java/org/apache/kyuubi/jdbc/hive/TestKyuubiPreparedStatement.java:74]: can we close the `resources`, e.g. connection, statement, result set, and remove such SuppressWarnings

Comment (turboFei) [reply kyuubi-hive-jdbc/src/test/java/org/apache/kyuubi/jdbc/hive/TestKyuubiPreparedStatement.java:74]: If it is needed, we can do it in another pr.

**Comment by turboFei** [comment]
thanks for your review

**Comment by yaooqinn** [comment]
please fill up all items in the sidebar

File: kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiStatement.java [hunk 10]
```diff
@@ -0,0 +1,975 @@
...
+
+  /*
+   * (non-Javadoc)
+   *
+   * @see java.sql.Wrapper#isWrapperFor(java.lang.Class)
+   */
+
+  @Override
+  public boolean isWrapperFor(Class<?> iface) throws SQLException {
+    return false;
```
Comment (SimpleExpress) [inline kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiStatement.java:847]: why return false without testing?

Comment (pan3793) [reply kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiStatement.java:847]: It's derived from Apache Hive, if you find some bugs that are not caused by Kyuubi modification, the perfect approach is to report and fix them in the upstream Hive, then port to Kyuubi.

Submitting to Kyuubi directly is also feasible if no response from the Hive community. 

https://github.com/apache/hive/blob/branch-2.3/jdbc/src/java/org/apache/hive/jdbc/HiveStatement.java#L860

**Pull request closed without merging**